File: modules/faithlog-kakao-share/android/src/main/java/expo/modules/faithlogkakaoshare/FaithLogKakaoShareModule.kt
package expo.modules.faithlogkakaoshare

import android.content.Intent
import android.content.pm.PackageManager
import com.kakao.sdk.common.KakaoSdk
import com.kakao.sdk.share.ShareClient
import com.kakao.sdk.template.model.Content
import com.kakao.sdk.template.model.FeedTemplate
import com.kakao.sdk.template.model.Link
import com.kakao.sdk.template.model.Social
import com.kakao.sdk.template.model.Button
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class FaithLogKakaoShareModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("FaithLogKakaoShare")
    AsyncFunction("share") { title: String, description: String, buttonTitle: String, url: String, promise: Promise ->
      val context = appContext.reactContext ?: run {
        promise.reject("KAKAO_CONTEXT_UNAVAILABLE", "Kakao share is unavailable.", null)
        return@AsyncFunction
      }
      val applicationInfo = context.packageManager.getApplicationInfo(
        context.packageName,
        PackageManager.GET_META_DATA
      )
      val appKey = applicationInfo.metaData?.getString("com.kakao.sdk.AppKey")
      if (appKey.isNullOrBlank()) {
        promise.reject("KAKAO_APP_KEY_MISSING", "Kakao share is unavailable.", null)
        return@AsyncFunction
      }
      KakaoSdk.init(context.applicationContext, appKey)
      val template = FeedTemplate(
        content = Content(title = title, description = description, imageUrl = "https://app.faithlog.kr/og-faithlog.png", link = Link(webUrl = url, mobileWebUrl = url)),
        social = Social(),
        buttons = listOf(Button(buttonTitle, Link(webUrl = url, mobileWebUrl = url)))
      )
      if (ShareClient.instance.isKakaoTalkSharingAvailable(context)) {
        ShareClient.instance.shareDefault(context, template) { sharingResult, error ->
          if (error != null || sharingResult == null) {
            promise.reject("KAKAO_SHARE_FAILED", "Kakao share failed.", error)
          } else {
            context.startActivity(sharingResult.intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
            promise.resolve("completed")
          }
        }
      } else {
        ShareClient.instance.shareDefault(context, template) { sharingResult, error ->
          if (error != null || sharingResult == null) {
            promise.reject("KAKAO_SHARE_FAILED", "Kakao share failed.", error)
          } else {
            context.startActivity(sharingResult.intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
            promise.resolve("completed")
          }
        }
      }
    }
  }
}

File: modules/faithlog-kakao-share/ios/FaithLogKakaoShareModule.swift
import KakaoSDKShare
import KakaoSDKTemplate
import KakaoSDKCommon
import React
import UIKit

@objc(FaithLogKakaoShare)
public final class FaithLogKakaoShareModule: NSObject {
  private static var initialized = false

  @objc public static func requiresMainQueueSetup() -> Bool {
    true
  }

  @objc public func share(
    _ title: String,
    description: String,
    buttonTitle: String,
    url: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let shareURL = URL(string: url),
          let imageURL = URL(string: "https://app.faithlog.kr/og-faithlog.png") else {
      reject("KAKAO_SHARE_INVALID_URL", "카카오톡 공유를 사용할 수 없습니다.", nil)
      return
    }
    guard let appKey = Bundle.main.object(forInfoDictionaryKey: "KAKAO_NATIVE_APP_KEY") as? String,
          !appKey.isEmpty else {
      reject("KAKAO_APP_KEY_MISSING", "카카오톡 공유를 사용할 수 없습니다.", nil)
      return
    }
    if !Self.initialized {
      KakaoSDK.initSDK(appKey: appKey)
      Self.initialized = true
    }
    let link = Link(webUrl: shareURL, mobileWebUrl: shareURL)
    let template = FeedTemplate(
      content: Content(title: title, imageUrl: imageURL, description: description, link: link),
      buttons: [Button(title: buttonTitle, link: link)]
    )
    ShareApi.shared.shareDefault(templatable: template) { result, error in
      guard error == nil, let result else {
        reject("KAKAO_SHARE_FAILED", "카카오톡 공유를 열지 못했습니다.", nil)
        return
      }
      DispatchQueue.main.async {
        UIApplication.shared.open(result.url, options: [:]) { opened in
          if opened {
            resolve("completed")
          } else {
            reject("KAKAO_SHARE_OPEN_FAILED", "카카오톡 공유를 열지 못했습니다.", nil)
          }
        }
      }
    }
  }
}
